source/_posts/2017-11-02-kafka-streams-dsl-vs-processor-api.markdown

```diff
@@ -0,0 +1,542 @@
+      } else {
+        pvs
+          .filter { pv =>
+            pv.pvId == ev.pvId
+          }
+          .map { pv =>
+            EvPv(ev.evId, ev.value, Some(pv.pvId), Some(pv.value))
+          }
+          .toSeq
+      }
+
+      evPvs.foreach { evPv =>
+        context().forward(evPvKey, evPv)
+        evPvStore.put(evPvKey, evPv)
+      }
+    }
+  }
+
+  private def isNotDuplicate(evPvKey: EvPvKey, timestamp: Long, deduplicationWindow: FiniteDuration) =
+    evPvStore.fetch(evPvKey, timestamp - deduplicationWindow.toMillis, timestamp).asScala.isEmpty
+
+  private def storedPvs(key: ClientKey, timestamp: Long, joinWindow: FiniteDuration) =
+    pvStore.fetch(key, timestamp - joinWindow.toMillis, timestamp).asScala.map(_.value)
+  }
+```
+
+First processor performs a lookup for previously joined `PvEv` by `PvEvKey`.
+If `PvEv` is found the processing is skip because `EvPv` has been already processed.
+
+Next, if page view does not exist for given event (`pvs.isEmpty`) processor returns event itself (left join semantics).
+
+If there are matched page views from given client, 
+try to match page view to event using simple filter `pv.pvId == ev.pvId`.
+We don't need any repartitioning to do that, only gets all page views from given client 
+and join with event in the processor itself. 
+It should be very efficient because every client generates up do hundred page views in 10 minutes.
+
+Perceptive reader noticed that processor also changes the key from `ClientId` to `EvPvKey`
+for deduplication purposes. Without any repartitioning, due to the fact that new key 
+is more detailed that original one. Everything is still within given client context. 
+
+As before windowed store for deduplication needs to be configured.
+Because deduplication is done in a very short window (10 seconds or so),
+the logging to backed internal Kafka topic is disabled at all.
+If one of the stream instance fails, we could get some duplicates during this short window, not a big deal.
+
+``` scala StoreBuilders
+val evPvStoreWindowDuration = 10 seconds
+
+val retention = evPvStoreWindowDuration.toMillis
+val window = evPvStoreWindowDuration.toMillis
+val segments = 3
+val retainDuplicates = false
+
+val evPvStore = Stores.windowStoreBuilder(
+  Stores.persistentWindowStore("ev-pv-window-store", retention, segments, window, retainDuplicates),
+  EvPvKeySerde,
+  EvPvSerde
+)
+```
+
+Add join processor to the topology and connect with event source upstream.
+
+``` scala
+val evJoinProcessor: ProcessorSupplier[ClientKey, Ev] =
+  () => new EvJoinProcessor("ev-join-processor", "pv-window-store", "ev-pv-window-store", pvStoreWindowDuration, evPvStoreWindowDuration)
+
+new Topology()
+  (...)
+  .addProcessor("ev-join-processor", evJoinProcessor, "ev-source")
+```
+
+The last processor maps compound key `EvPvKey` again into `ClientId`. 
+Because client identifier is part of the compound key, repartitioning is done by processor without repartitioning.
```

Comment:
nie wiem czy dobrze zrozumiałem, ale napisałbym tak:
Because client identifier is part of the compound key it is repartitioned already without the need for further repartitioning.